### DISCUSSION TITLE
Merging HsExpr, HsType, and Pat

### DISCUSSION THREAD

**Discussion by goldfirere** [ghc-proposals/ghc-proposals#663] [General]
This is a discussion on changing internal implementation details of GHC. It is *not* a proposal for a user-facing language change. However, because many various tools depend on the internals of GHC, it seems appropriate to open this up to the community for feedback. But the internal nature of this is why this is a GitHub discussion and not a proper pull request. The ideas here arose in a conversation between @simonpj, Artin Ghavisand (whose handle I don't have), and me.

# Overview

We think it might be good to merge the `HsExpr`, `HsType`, and `Pat` types in GHC.

# Background

GHC has a large collection of datatypes used to hold its abstract syntax. Prominent among these are `HsExpr`, for expressions, `HsType`, for types, and `Pat`, for patterns. These three have various bits already in common:
* All support variables: `HsVar`, `HsTyVar`, and `VarPat`
* All support application: `HsApp`, `HsAppTy`, and `ConPat` (though this is n-ary)
* All support type application: `HsAppType`, `HsAppKindTy`, and `ConPat` (again, n-ary)
* All support tuples (boxed and unboxed), lists, literals, infix application, unboxed sums, parentheses, and type signatures

Furthermore, patterns are already parsed as expressions and only later converted from `HsExpr` to `Pat`.

# Proposal

Phase 1:

1. Mint a new type `HsTerm` that contains all of the syntactic constructs currently stored in all of `HsExpr`, `HsType`, and `Pat`, deleting these other types in favor of the new one.
2. Combine the pretty-printers for these types.
3. Remove the current `HsExpr`-to-`Pat` conversion function.
4. Keep the parsers distinct (for now)
5. Keep the name resolution for patterns separate: it brings variables into scope. Combine the name resolution algorithms for expressions and types, tracking the selection of namespace in a little context argument passed to recursive calls of the function.
6. Keep type-checking & desugaring separate.

Phase 2:

7. Combine the parsers. If we fold in `HsCmd` (the syntax used for commands in `Arrows`-syntax), then we can get rid of the uncomfortable `DisambECP` construct. This will also require changing `RdrName` not to carry a namespace, but that seems like an improvement: the parser has no business figuring out the namespace of identifiers. That's the renamer's job.

# Rationale

Simplification. We think this will work out nicely, based on work we're doing writing up a formalism of the rules type-checking the surface language. Those rules prefer having a unified syntax between these forms, and this suggests that the implementation will follow suit.

This will all be an experiment. If it turns out that this simplification isn't possible, then we can abandon the experiment.

**Comment by int-index** [comment]
> Furthermore, patterns are already parsed as expressions and only later converted from `HsExpr` to `Pat`.

We don't convert from `HsExpr` to `Pat`. We used to do it before https://github.com/ghc/ghc/commit/52fc2719b93ab39be3e52eba531ee173b9134183 (2019), but not anymore. The whole point of `DisambECP` is to avoid such a conversion.

So...

> Remove the current HsExpr-to-Pat conversion function.

There's no such function.

**Reply by Ericson2314** [reply]
That's definitely an improvement, but still fits in the spirit of the above, right? E.g. there are type patterns which may or may not be in `HsPat` today (I forget), and now they would be for sure: only `HsExpr` and `HsPat` remain.

**Reply by phadej** [reply]
patterns are "binders", expressions and types are "terms". Don't mix them. Having `Expr` (e.g. let-expression, or any redux) on the LHS of e.g. let binding doesn't make sense.

**Reply by Ericson2314** [reply]
@phadej I am agreeing let's keep `HsExpr` and `HsPat` separate. I agree that @int-index's solution is better, and obviates the need to remove `HsPat`.

**Comment by Ei30metry** [comment]
> This is a discussion on changing internal implementation details of GHC. It is _not_ a proposal for a user-facing language change. However, because many various tools depend on the internals of GHC, it seems appropriate to open this up to the community for feedback. But the internal nature of this is why this is a GitHub discussion and not a proper pull request. The ideas here arose in a conversation between @simonpj, Artin Ghasivand (whose handle I don't have), and me.

Thanks for mentioning me Richard! I always use `Ei30metry` as my handle.

**Comment by Ericson2314** [comment]
Yay it would be very good if the type---term merge finally happens! (The `HsExpr` <- `HsType` part, I agree that @int-index's solution for `HsPat` is good and those should stay separate.)

(CC @alt-romes, if the AST split-out happens first (with a good single `GHCi` / `HLS` development experience of course), I am curious how that would affect the visibility of this momentous occasion :)).

**Comment by phadej** [comment]
I don't like this idea. As long as types, expressions and patterns are syntactically distinct (and patterns vs. types/terms will stay so), it's really better for the codebase, even that it feels that at a moment there is some repetition.

Consider TH translation. The `HsExpr` is converted into TH-combinators. Would TH interface change to? Will the conversion function become partial?

**Reply by Ericson2314** [reply]
Oh I guess Richard did call out the `DisambECP` construct --- yes I agree that keeping patterns and type/expressions separate is good. Just the `HsType` and `HsExpr` merge I am excited about. Edited my comment to make that clearer.

**Comment by goldfirere** [comment]
Combining the datatypes says nothing about the algorithms that work over those datatypes. So there needn't be any impact, at all, on TH.

I find the proposal here considerably simpler than today's `DisambECP`. No fancy technology! Just a datatype.

**Reply by Ericson2314** [reply]
Eh it's "fancy tech" (not cumbersome is it? I wouldn't think so) in one module vs nicer pattern matching in every other module? Seems like a good trade-off to me!

The things that make GHC hard, at least for me, is when some phenomenon is an emergent property of a bunch of different code scattered around. Something like `DisambECP` which is appears to be very locally scoped strikes me as quite harmless.

**Comment by sgraf812** [comment]
I've not worked with the frontend types nearly as much as other people in this thread, but I feel quite sympathetic to the proposal.

The main benefit is that we get *one* data type definition which allows us to share code.
Still, there is value in being able to match exhaustively on a pattern (for example).
In the current implementation, the `DisamECP` mechanism disambiguates and validates a syntactic expression into a pattern, and we pass around `HsPat`s ever after as a witness of that validation. If we merge `HsExpr` and `HsPat`, shouldn't we find a way to keep around the distinction between post-validation as well? 

I haven't worked this out completely, but perhaps we can have a solution based on TTG? E.g.
```hs
type data ExprPhase p = Expr p | Pat p
type HsExpr p = HsTerm (Expr p)
type HsPat  p = HsTerm (Pat p)

data HsTerm p = ...
  | OrPat (XOrPat p) ...
  | HsDo (XDo p) ...

type instance XDo            (Expr GhcPs) = AnnList
type instance XDo            (Expr GhcRn) = NoExtField
type instance XDo            (Expr GhcTc) = Type
type instance XDo            (Pat GhcPs) = AnnList
type instance XDo            (Pat _) = DataConCantHappen

type instance XOrPat (Pat GhcPs) = NoExtField
type instance XOrPat (Pat GhcRn) = NoExtField
type instance XOrPat (Pat GhcTc) = Type
type instance XOrPat (Expr GhcPs) = NoExtField
type instance XOrPat (Expr _) = DataConCantHappen
```

Note that `Expr GhcPs` must be fairly lenient since validation/disambiguation happens post parsing.

**Reply by Ericson2314** [reply]
I think we should first be clear on what problems `DisambECP` has that we're trying to fix. I don't think it has any problems whatsoever.

A TTG solution would leak the existence of parsing problem into the AST data definitions. Conversely, `DisambECP` scopes the parsing problem to just the parser, so no other code need be aware of it.

**Reply by sgraf812** [reply]
> A TTG solution would leak the existence of parsing problem into the AST data definitions.

I see no downsides to that. Other languages do not even have a concept of a closed-union syntax tree pre elaboration (e.g. Lean), just S-expressions, and they do fine. (NB: It's hardly acceptable for today's LSP tooling to *just parse*. Elaboration is mandatory.)

> I think we should first be clear on what problems DisambECP has that we're trying to fix

DisambECP does two things:
1. It delays the decision of whether we need some term as Expr, Pat or Command for later *disambiguation*
2. Once the term can be disambiguated, it *validates* the encoded term (e.g. no `OrPat` in `HsExpr`, no `HsDo` in `HsPat`), returning an `HsExpr`, `HsPat` or `HsCmd`

Whatever we do, we need to keep doing (2). The OP would just do so at different types and perhaps delay validation to the renamer.

Regarding (1), note that `DisambECP` is "simply" a final encoding of the proposed `HsTerm`.
We could make it an initial encoding today (would be simpler but perhaps slower), at which point (2) is a triple of functions `checkExpr :: HsTerm -> PV (HsExpr GhcPs)`, `checkPat :: HsTerm -> PV (HsPat GhcPs)` and another one for commands.

The OP would reasonably simplify this setup, because the `check` functions could essentially be validation in `RnM` + `coerce`.
A reasonable simplification.

On a more architectural level, I dislike doing (2) post-parsing because the code is *entirely separate* from the grammatical specification which is the .y file.
The principled solution (one I would pursue given infinite dev resources) would be to switch to a performant GLR parsing backend for `happy` that disambiguates patterns from expressions *as expressed in the .y file*! Thus the .y file is *the* specification for what is parsed.
No need for (2) in this world.

**Reply by Ericson2314** [reply]
@sgraf812 

I too would have no problem switching from a "final" to "initial" way of doing it, so long as we still had `HsExpr` and `HsPat` to *prove* validation has occurred (a la "parse not validate", "make illegal states unrepresentable" etc.).

But proposal is to have an `HsTerm` *and get rid of* `HsExpr`, `HsPat` --- and I think the "getting rid of" part is the more important thing for @goldfirere than initial vs final in the parser. Throwing away the proof so all other code has incomplete patterns is what feels very wrong to me.

> The principled solution (one I would pursue given infinite dev resources) would be to switch to a performant GLR parsing backend for happy that disambiguates patterns from expressions as expressed in the .y file! Thus the .y file is the specification for what is parsed.

Yes I agree that is good. (It does leave less opportunity for domain-specific errors, but with even more formal language research we can hopefully get those back too :).)

**Reply by sgraf812** [reply]
> so long as we still had HsExpr and HsPat to prove validation has occurred (a la "parse not validate", "make illegal states unrepresentable" etc.).

Do note that my idea gives you exactly that. After validating that a `HsTerm ` can be a Pat, the validation function (which I did not call `parse` because it would have been confusing) returns the `HsTerm (Pat GhcPs)`, which is just `HsPat GhcPs`. The implementation can simply be `coerce` (perhaps an unsafe one) once we have validated. 

> But proposal is to have an HsTerm and get rid of HsExpr, HsPat --- and I think the "getting rid of" part is the more important thing for @goldfirere than initial vs final in the parser.

Well, that is ultimately for Richard to clarify, but my impression is he wasn't proposing to use unvalidated `HsTerm` all over the place.

Hence I found it important to state that we can have *both* code sharing on a unified `HsTerm` *and* good validation, and that doing one does not rule out the other.

**Comment by simonpj** [comment]
Some thoughts:

* This proposal is purely an implementation proposal.  It should have no impact on
  the language as seen by the programmer, or on Template Haskell.

* At *some* point we have to validate whether, say `forall a. a->a` is OK at a particular place.  Traditionally we do that in the parser, but there are plenty of places where we delay, parsing a more accommodating grammar and sorting it out later.  Reason for moving error reporting out of the parser: 
  * It can be hard to figure out what is OK in the parser (e.g. precendence parsing errors), 
  * We get much better error messages, because in the renamer/typechecker we have much more context to report
  * We get much better error recovery.  Typically parse errors are fatal, whereas renamer/typechecker errors can often recover and thus report multiple errors in one run.  This matters in HLS.

  This proposal suggests doing more of this: parse into an accommodating `HsTerm`, and validate later when we have more context to be able to report good errors.

* Dependent Haskell has forced us to include all of the syntax of types (usually `HsType`) in the syntax of `HsExpr`, and sort it out later, so that we can accommodate `RequiredTypeArguments`; e.g. `f (forall a. a->a)` has no parser-visible clue to switch to `HsType`, so we parse it all as `HsExpr`.  We sort it out later, in the type checker.

  So if we see `f (case x of ...)` where `f :: forall a -> blah`, we'll get a helpful error (during typechecking!) that you can't have a `case` expression in a type.

  It would be fantastic to be able to drop all the "term-to-type" stuff in DH. That's exactly what the formal typing rules we are writing do, which is where this proposal sprang from.

* Separating `HsType` and `HsExpr` forces duplication in the renamer.  Renaming types and terms is (explicitly by design!) practically identical, except for which is the "default namespace" to start with when looking up an identifier.

  If we abandoned `HsType` entirely, we'd share that code.  In exchange, we'd get a slightly later error if we write
  ```
  f :: forall a. case a of ...
  ```
  Instead of being rejected by the parser, we'll get a nicer error message from the renamer or typechecker saying "you can't have a case expression in a type".  Just like `RequiredTypeArguments` above.

* The `DisambigECP` stuff is extremely ingenious, but it is also quite complicated. Vlad kindly explained it to me, but I have since forgotten how it works.  It would be *much* easier simply to parse into a single data type.

* Patterns are a bit less clear-cut than types, for two reasons:
  * the renamer for patterns is not the same as the renamer for terms, so we don't get any sharing there
  * the curried form of application is not helpful for patterns, which need to see fully-saturated constructor applications.

  My instinct is that (unlike `HsType` which we could abandon entirely) we might retain `HsPat`, but have
  ```
  rnPat :: HsTerm GhcPs -> RnM (HsPat GhcRn)
  ```
  That is, the pattern renamer validates that the `HsTerm` is indeed a pattern, and turns it into a `HsPat`.

To me this looks very attractive:
* Fewer data types
* Much simpler parser
* Better error meessages
* Less code (no term-to-type conversion)
* More direct expression of [our design principles](https://github.com/ghc-proposals/ghc-proposals/blob/master/principles.rst#2language-design-principles)

I'm not sure how all this affects `HsCmd`, but I expect it'd be similar.

**Comment by maralorn** [comment]
Side note: This should probably not be an argument because there are also better fixes planned for this issue, but as it currently stands moving errors out of the parser into the type checker improves usability e.g. in hls a lot, because those errors then don’t prevent other type errors elsewhere in the file from being shown.

**Reply by simonpj** [reply]
> Side note: This should probably not be an argument because there are also better fixes planned for this issue, but as it currently stands moving errors out of the parser into the type checker improves usability e.g. in hls a lot, because those errors then don’t prevent other type errors elsewhere in the file from being shown.

Can you elaborate?  What are the "better fixes planned"? 

I'll edit my post to mention that moving errors out of the parser also improves error recovery -- good point.

**Reply by maralorn** [reply]
I don‘t know the details and how far this is but at the ecosystem workshop people where discussing giving the parser the ability to still parse other declarations when it fails on some declarations.

**Reply by googleson78** [reply]
I believe this is the related work for allowing multiple parse errors - https://github.com/haskell/happy/pull/272

**Reply by sgraf812** [reply]
Indeed, @Kariiem is our GSoC student and is currently hard at work merging https://github.com/haskell/happy/pull/272 bit by bit. This is part of a larger tech proposal by @michaelpj here: https://github.com/haskellfoundation/tech-proposals/blob/f79c2940f36e05bf587fb6b9b7ee0dce09f041f0/proposals/0000-fault-tolerant-ghc.md

However, I don't think that the work on `happy` affects the parse validation errors that are the subject of this discussion.
They are non-fatal *today*, so it should be entirely possible to resume parsing (I think it's likely we do already). However, we do not currently emit renaming or type checking errors after a non-fatal parsing error. The reason for that is simply that we do not have the concept of "error nodes", as it is called in the tech proposal.

I think introducing those error nodes is far simpler than the refactoring proposed in the OP.

So: TLDR; I think "Better error meessages" is a red herring and IMO not a good reason we should be pursuing this proposal.

**Reply by maralorn** [reply]
Yeah, I assumed that much. But thank you for clarifying!

**Comment by s-and-witch** [comment]
Hello there! 

I am glad that the other people working closely on the GHC also feel the ripe opportunity to unify existing data types.

I have the patch on the fly [1] that would make `HsType` almost a subset of `HsExpr` with a few exceptions [2]. This patch is another step forward in iterative merging `HsType` with `HsExpr`. 

However, I don't see a good reason for merging `HsExpr` with `HsPat`. 

## Issue statement

`HsExpr` contains plenty of nodes that possibly may occur in types someday in a happy future, but are meaningless in patterns (let, case, label, lambda, and so on), and `HsPat` also contains pattern-only nodes (or-pat, as-pat).

The main possible reason for unifying data types is to share common implementation details. However, renaming and type-checking for patterns and expressions would stay distinct, and the parser is already unified. So, what's left besides pretty-printing? There's not too much code IMO.

## Possible solutions

We can rest huge ~~God object~~ `HsTerm` data type with all the constructors included and cut off invalid parts at runtime. That would be a strike for all the clients of GHC API that want to perform static analysis on expressions because they would have to perform the same cut-offs themselves.

And this brings us to another part of this problem: API discoverability.  `HsType`, `HsExpr`, and `HsPat` say all about valid syntactic forms of patterns, expressions, and types, but with them all merged, the source of truth would be in the internals of validation functions (I suppose that would be during renaming).

I would say that a few more specialized data types may be simpler than one data type that combines them but has more hidden invariants inside.

---

It is possible to encode this distinction in a single `HsTerm` via introducing a new dimension in extension points, like what @sgraf812 proposed here https://github.com/ghc-proposals/ghc-proposals/discussions/663#discussioncomment-9917283. I don't like this complication that will (unlike `DisambECP`, which lives in the parser solely!) infect the whole compiler frontend.

However, the distinction between `HsPat` and `HsExpr` cannot be expressed by banning unwanted nodes, as this code example shows: `(f g) h` is a valid expression, but an invalid pattern. 

It is possible to express this invariant statically, but it requires an additional type family wrapper for the application head. Even more complexity for the change that aims to simplify things!

This design provides better API discoverability than the first solution but worse than the current state: a newcomer would have to seek for type family instances (which is painful by itself, yet better than seeking required bits in the source code of a function that does a lot of additional things). 

Keep in mind that we want to move Haskell syntax from GHC into a dedicated library: I don't know any client that doesn't want to have a distinction between expression and pattern, so `data ExprPhase p = Expr p | Pat p` should live in this library, with all that `DataConCantHappen` stuff.

---

@simonpj proposed this function to cut off invalid nodes from patterns (https://github.com/ghc-proposals/ghc-proposals/discussions/663#discussioncomment-9922980):
```
rnPat :: HsTerm GhcPs -> RnM (HsPat GhcRn)
```

Having such API means that all pattern stuff will live in `HsTerm` with all the same problems I wrote above. 

One of the goals that we want to achieve is:
> Remove the current HsExpr-to-Pat conversion function.

As @int-index mentioned above (https://github.com/ghc-proposals/ghc-proposals/discussions/663#discussioncomment-9909493) we already don't have this function, but having such `rnPat` means a step backward to the previous state.

## Summary

- I really, really like to have `HsType` and `HsExpr` merged. I like this idea so much that _I am already taking steps forward in this direction_ [1]

- I doubt that merging `HsExpr` with `HsPat` is an improvement. 

---
[1] https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12910
[2] There still are -`XStarIsKind`, `HsBangTy`, and `HsRecTy` to be resolved before merging data types (the latter one probably just should be excluded from the `HsType`).

**Reply by sgraf812** [reply]
> I doubt that merging HsExpr with HsPat is an improvement.

FWIW, I tend to think the same. Above I simply wanted to show that it's not fundamentally impossible to merge the `HsPat` data type into `HsTerm` as well, and that it does not mean we have to give up on static guarantees post validation.

Ultimately, I think the question of whether or not we can get rid of `DisambECP` in favor of a simpler, initial encoding is entirely independent of the bigger goal: Unifying `HsExpr` and `HsType`. Thus the following is perhaps off-topic, and the OP would perhaps be less contentious if it simply left out the `DisambECP` mechanism.

---

I think that an initial encoding of `ECP` which we could call a `PreTerm`, comprising the syntax of `HsExpr`, `HsPat` and `HsCmd`, would be simpler to understand. I don't know whether Vlad benchmarked at the time to see whether the perf difference is noticable. The argument "we build the AST twice" doesn't really convince me, because the final encoding has practically the same footprint.
However I do not think it's good to spend our time here; what we have is a bit too fancy for my tastes, but it works well and is isolated.

I'm not optimistic that Simon's idea `rnPat :: HsTerm GhcPs -> RnM (HsPat GhcRn)` would work well, because then the type of e.g., the `m_pats` field of `Match` would need to change from `HsTerm GhcPs` to `HsPat GhcRn` when going from `GhcPs` to `GhcRn`. Well, not without severly complicating the definition of `HsPat p`, at least.

**Reply by s-and-witch** [reply]
As I said above, I'm iteratively working under unification of `HsType` and `HsExpr`. I don't think that this change require some The Grand Patch that does everything in one transaction. Here are a few steps that we should to do before we can have `type HsType = HsExpr`:

- Merge https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12910
- Get rid of `HsRecTy` (this work probably belongs to https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0402-gadt-syntax.rst)
- Add `instance DisambECP (HsType GhcPs)`
- Add missing bits of `HsType` into `HsExpr` (`StarKind`, `HsBangTy`)
- Something else that I don't remember

And after doing those points would be possible to have patch with
```
type HsType = HsExpr
```
And substitution of `HsType` with `HsExpr` in a few functions where the logic for type and expression differs, like `cvtType`, `tcHsType`.

**Reply by phadej** [reply]
`StarKind` is `* = Type :: Type` isn't it? That syntax is marked as deprecated. I'd rather remove it all-together (and e.g. make parser produce the `Type`). Something like `id * Bool` won't ever mean `id Type Bool`; and pretty-printing should never produce that either, even if `id StarKind Bool` like `HsExpr` would make sense internally.

`HsBangTy` is also a bad construct. It's something which cannot appear in type annotations nor expressions. It's a syntax of data-declarations. I guess it's there so the types in GADTs syntax can be "easily" parsed. But this "easily" is not correct. The `HsBangTy`, or `UNPACK` pragma don't belong to expression-types.

TL;DR, clean up warts first.

**Reply by s-and-witch** [reply]
That's all true. 
> I'd rather remove it all-together (and e.g. make parser produce the Type). 

Still, the parser has to be able to deal with *. There is a WIP branch that attempts to workaround it, by allowing * kind to occur only around `->` and at the top level (That should be enough to express `Haskell98` kinds).

About `HsBangTy` let's substitute my previous "Add missing bits of `HsType` into `HsExpr`" with "we have to figure out some solution, like for `HsRecTy`, because they caused by the same problem."

**Comment by int-index** [comment]
> It would be fantastic to be able to drop all the "term-to-type" stuff in DH.

Yes, I agree. With DH, the case for unifying `HsExpr` and `HsType` is very strong, for two reasons.

1. We already want to pass types as function arguments (e.g. with `RequiredTypeArguments`), so type syntax such as `Int -> Bool` or `forall a. Num a => a` needs to be included in `HsExpr`.
2. In the future, we also hope to support arbitrary computation at the type level, so various constructs such as `case`, `let`, `if`, etc, would have to be included in `HsType`.

In other words, constructs that were exclusive to `HsType` are making their way into `HsExpr`, and stuff exclusive to `HsExpr` will have to be added to `HsType`. The syntax is "leaking" in both directions, and the end goal is to have no distinction between those two at all.

However! The situation is very different with `HsCmd` and `Pat`. Here are some constructs that are exclusive to patterns and it'd make no sense to bring them to expressions:

* as-patterns `x@p`
* view-patterns `e -> p`
* lazy patterns `~p`
* bang patterns `!p`
* or-patterns `p1; p2`

We simply shouldn't have those when we are dealing with expressions. Similarly, here are the constructs that are exclusive to commands:

* `f -< arg`
* `arg >- f`
* `f -<< arg`
* `arg >>- f`
* `(| e cmd1 .. cmdn |)`

Those are simply not valid expressions, and there's nothing to suggest that they should be.

I am not in principle opposed to finding a simpler alternative to `DisambECP`, and perhaps this alternative is indeed throwing `HsExpr`, `HsType`, `HsCmd`, and `Pat` into a blender and using the resulting slop as our main data type in the GHC front-end. I just want to point out that this is quite different from unifying `HsExpr` and `HsType`, which is a way more principled move reflecting actual developments in language design (with DH we want to treat types as values and use values in types).

**Comment by phadej** [comment]
> . With DH, the case for unifying HsExpr and HsType is very strong

It would be *stronger*, if `Type` and `CoreExpr` were the same (or at least `Type` would be a true subtype of `CoreExpr`). They aren't.

I.e. I'd prefer moving towards DH from the "backend" of the compiler, not from the frontend. Even if we unify the `HsExpr` and `HsType`, at some point we need to split those into `CoreExpr` and `Type`.

On the other hand that split would probably happen in type-checker, which reminds me about another issue: the `NoGhcTc` type family. That is a GHC specific type family, which is defined in "generic" `Language.Haskell.Syntax` namespace. That is consequence of `HsType GhcTc` not really existing, at that point the types in the AST are already Core `Type`s (and are in TTG extension points).

How the unification of `HsExpr` and `HsType` will affect that? If unification will allow us to get rid of `NoGhcTc`, great. If it will making removal even harder, than it's not good.

**Reply by int-index** [reply]
Yes, I agree with you. The Core language needs to be updated as well, and I'm not a fan of `NoGhcTc` either. Just trying to do one thing at a time.

**Reply by phadej** [reply]
I'm arguing about an order in which things could be done. No need to do everything at once.

IMHO, if we cannot update Core to be DH-ideas-compatible, there is no reason to update frontend either. All the risk of success vs. failure lies whether we can make Core stuff work. (Core definitions, optimisations, translation to STG). The frontend will then follow more naturally.

**Reply by simonpj** [reply]
> IMHO, if we cannot update Core to be DH-ideas-compatible, there is no reason to update frontend either. All the risk of success vs. failure lies whether we can make Core stuff work. (Core definitions, optimisations, translation to STG). The frontend will then follow more naturally.

I respectfully disagree.  I think there are big gains to be had in the front end (see above) entirely independently of whether Core has terms and type combined.  The latter would be a huge step with profound consequences for Core, and major costs.  The former is a relatively simple refactoring, with no consequences for Core.

**Reply by phadej** [reply]
> The former is a relatively simple refactoring,

Not necessarily to all the tools using GHC frontend. Every change to `HsExpr` is a cost. I stopped updating GHC plugins I don't use actively, because the accumulated cost become too big. (To be honest, I stopped using GHC source plugins all together, as I consider them a huge risk from the maintenance and upgrading point-of-view).

**Reply by Ericson2314** [reply]
> (To be honest, I stopped using GHC source plugins all together, as I consider them a huge risk from the maintenance and upgrading point-of-view)

(Off topic, but I think this is the correct view, and also not a problem. I agree with @lexi-lambda that a rejuvenated TH with a better compat story is the right way to solve these problems.)

**Comment by simonpj** [comment]
There seems to be some consensus on abolishing `HsType` in favour of `HsTerm`.  I'm content with that as a first step.   We could do that, and evaluate.

But I'm not convinced of the arguments for not doing the same with `HsPat`.   I agree that not much code-sharing is obtained, because patterns are renamed quite differently to types.  But

* Yes, as-patterns (say) can occur only in patterns. But whether we reject an as-pattern appearing in a term (a) in the parser or (b) in the renamer, seems a relatively minor consideration.

* The DisambigECP mechanism is, as I say above, clever but complicated. I'd love to make it unnecessary.

An alternative I had not previously thought of would be this: use `HsTerm` for both `HsExpr` and `HsType`; and use `HsPat` for both the current `HsPat` and for type patterns.  Meta-goal: eliminate pat-to-type as well as term-to-type.

**Reply by Ericson2314** [reply]
@simonpj Yes that's the alternative I think many of us want: save the pattern stuff for later, and still get rid of the other term<->type headache.

We can always reconsider unifying `HsTerm` and the new improved `HsPat` (for both expr-patterns and type-patterns) later.

**Reply by Ericson2314** [reply]
> But whether we reject an as-pattern appearing in a term (a) in the parser or (b) in the renamer, seems a relatively minor consideration.

> The DisambigECP mechanism is, as I say above, clever but complicated. I'd love to make it unnecessary.

For both of those points, I would re-read what @sgraf812 is saying:

- `DisambigECP` has methods which are 1-1 with `HsTermAndPat`'s constructors

- We can convert `HsTermAndPat` to `HsTerm` and `HsPat` in a simple pass *prior* to renaming (better separation of concerns)

- We don't have to emit fatal errors, can instead e.g. emit warnings and replace invalid syntax with with `_` patterns, for sake of HLS etc. Therefore there should be no impact on errors regardless.
